mlx/backend/cpu/encoder.h

```
@@ -0,0 +1,50 @@
+// Copyright © 2025 Apple Inc.
+
+#pragma once
+
+#include <unordered_map>
+
+#include "mlx/array.h"
+#include "mlx/scheduler.h"
+
+namespace mlx::core::cpu {
+
+struct CommandEncoder {
+  CommandEncoder(Stream stream) : stream_(stream) {}
+
+  CommandEncoder(const CommandEncoder&) = delete;
+  CommandEncoder& operator=(const CommandEncoder&) = delete;
+  CommandEncoder(CommandEncoder&&) = delete;
+  CommandEncoder& operator=(CommandEncoder&&) = delete;
+
+  void set_input_array(const array& a) {}
+  void set_output_array(array& a) {}
+
+  // Hold onto a temporary until any already scheduled tasks which use it as
+  // an input are complete.
+  void add_temporary(array a) {
+    scheduler::enqueue(stream_, [a = std::move(a)]() {});
+  }
+
+  // Hold onto temporaries until any already scheduled tasks which use it as
+  // an input are complete.
+  void add_temporaries(std::vector<array> arrays) {
+    if (arrays.empty()) {
+      return;
+    }
+    scheduler::enqueue(stream_, [arrays = std::move(arrays)]() {});
+  }
```

Not a problem necessarily but `add_temporaries` on the CPU "CommandEncoder" has to be after `dispatch` while on the GPU it can be before. One solution would be for the `CommandEncoder` to collect temporaries and then `cpu::eval` to dispatch the task that keeps them in memory possibly together with the rest of the buffers.

This would require `cpu::eval` to know about the `CommandEncoder` but it is already the case in the metal side so I don't think this is a problem and it will allow us to use `add_temporary` where it makes intuitive sense and reduce the number of tasks in the queue.

---

I think it's a very good suggestion as it's actually really not good that `add_temporary` has to be called at just the right spot and otherwise it could fail silently.